---
title: "Supplemental Figure 3: Carbapenem resistance-associated genotype matrix"
author: "Kyle Gontjes"
date: "`r Sys.Date()`"
output: html_document
---

```{r setup, include=FALSE}
require(knitr)
opts_knit$set(echo = F,warning = FALSE, message = FALSE,error=F,comment=NA) 
opts_chunk$set(echo = F,warning = FALSE, message = FALSE,error=F,comment=NA)
opts_knit$set(root.dir='~/Desktop/gl_nfs/Project_INTEGRATE/Analysis/Resistance_dynamics/INTEGRATE-resistance-ms')
```

# Environment

```{r}
# Packages
packages <- c("tidyverse","gridExtra","kableExtra","tableone","gontjesr","cowplot","ComplexHeatmap",'ggalluvial')
lapply(packages,library,character.only=T)

# Functions
source("./lib/consistent_themes.R")

# Print environment
Sys.info()
sessionInfo()
```

# Load data

```{r}
df <- readRDS("./data/dataset/CA_LTACH_first_clinical_CRKP_genomes_merged_studies_metadata.RDS")   
   
df_201415 <- df%>% subset(study == "2014-15")
df_202123 <- df%>% subset(study == "2021-23") 
```

# Coordinate ST data

```{r}
OLD_ST <- unique(df_201415$Species_ST) %>% sort
new_ST <- unique(df_202123$Species_ST) %>% sort 
ST_over_1 <- table(df$Species_ST) %>% subset(.>1) %>% names

df$Species_ST_new <- ifelse(df$Species_ST %in% c(ST_over_1),df$Species_ST,"Singleton STs") %>% gsub("Klebsiella pneumoniae ","",.)
                                          
break_of_sts <-  c(sort(unique(df$Species_ST_new %>% subset(.!="Singleton STs"))),"Singleton STs")
break_of_sts_renamed <- gsub("Klebsiella pneumoniae ","",break_of_sts)
       
ST_scale <- scale_fill_manual(breaks = break_of_sts,values = c(hues::iwanthue(n = length(unique(df$Species_ST_new)) -1),"lightgray"),labels = break_of_sts_renamed,name = "Sequence type", guide = guide_legend(ncol=7, title.position = "top", label.position = "right"))

df$Species_ST_new_no_species <- gsub("Klebsiella pneumoniae ","",df$Species_ST)
df$major_STs <- ifelse(df$ST %in% c("258","14","307","17","25","11"),df$Species_ST %>% gsub("Klebsiella pneumoniae ","",.),"Other STs")
```

```{r}
amrfinder <- readRDS("./data/amrfinder/amrfinder_curated.RDS")
amrfinder <- subset(amrfinder,amrfinder$isolate_no %in% df$isolate_no) 
amrfinder_beta <- amrfinder %>% subset(class == "BETA-LACTAM")
amrfinder_beta_bla <- amrfinder %>% subset(class == "BETA-LACTAM"& grepl("bla",element_symbol))
 
amrfinder_carb_bla <- amrfinder %>% subset(subclass == "CARBAPENEM" & grepl("bla",element_symbol) | grepl("KPC|NDM|VIM|IMP",element_symbol) | grepl("OXA-48 family",element_name))
amrfinder_carb_ompK <- amrfinder %>% subset(subclass == "CARBAPENEM" & ! element_symbol %in% amrfinder_carb_bla$element_symbol)
amrfinder_matrix <- readRDS("./data/amrfinder/amrfinder_matrix.RDS")  %>% subset(rownames(.) %in% df$isolate_no) 
amrfinder_matrix_beta <- amrfinder_matrix %>% select(any_of(sort(c(unique(amrfinder_beta$element_symbol)))))
amrfinder_matrix_carb <- amrfinder_matrix %>% select(any_of(sort(c(unique(amrfinder_carb_bla$element_symbol)))))

# Get an isolate's carbapenemase genes
df$blactamase <- sapply(df$isolate_no,FUN=function(x){
  amrfinder_beta_bla %>% subset(isolate_no == x) %>% .$element_symbol %>% paste0(collapse=" & ")
}) %>% {ifelse(.=="","No blactamase",.)}

df$carbapenemase <- sapply(df$isolate_no,FUN=function(x){
  amrfinder_carb_bla %>% subset(isolate_no == x) %>% .$element_symbol %>% paste0(collapse=" & ")
}) %>% {ifelse(.=="","No carbapenemase",.)}

df$carbapenemase_family <- sapply(df$isolate_no,FUN=function(x){
    carbapenemase <- amrfinder_carb_bla %>% subset(isolate_no == x) %>% .$element_symbol
     if(is_empty(carbapenemase)){
      "No carbapenemase"
    } else {
      sapply(carbapenemase,FUN=function(y){
        str_split(y,"-",simplify=T) %>% .[,1]
    }) %>% unlist  %>% paste0(collapse=" & ") 
    }
})     

df_201415 <- df%>% subset(study == "2014-15")
df_202123<- df%>% subset(study == "2021-23")
```

# Supplemental Figure:
```{r,fig.width=9.5,height=6.7857}
kleborate_matrix <- readRDS("./data/kleborate/kleborate_kpsc_both_studies_matrix.RDS") %>% select_if(!colnames(.) %in% c("OmpK36GD","OmpK36TD"))
df <- left_join(df,kleborate_matrix %>% mutate(isolate_no = rownames(.))) %>% left_join(amrfinder_matrix_beta %>% mutate(isolate_no = rownames(.))) %>% as.data.frame
rownames(df) <- df$isolate_no
df$major_STs <- ifelse(df$ST %in% c("ST258","ST14","ST307","ST17","ST25","ST11"),df$Species_ST %>% gsub("Klebsiella pneumoniae ","",.),"Other STs")
df_carb_omp <- df %>% select(study,carbapenemase,major_STs,any_of(c(unique(amrfinder_beta$element_symbol) ,colnames(kleborate_matrix)))) 
```

```{r}
# Omp mutations
kleb_matrix_count <- df_carb_omp %>% select(colnames(kleborate_matrix)) %>% colSums() %>% sort
mutations_over_4 <- names(subset(kleb_matrix_count,kleb_matrix_count>4))
muts_under <- subset(names(kleb_matrix_count),!names(kleb_matrix_count) %in% mutations_over_4)
ompK35_truncations_under_4 <- subset(muts_under,grepl("K35-",muts_under))
ompK36_truncations_under_4 <- subset(muts_under,grepl("K36-",muts_under))

df_carb_omp$ompK35_rare_truncation <- df_carb_omp %>% select(ompK35_truncations_under_4) %>% rowSums(.) %>% {ifelse(.>0,1,0)}
df_carb_omp$ompK36_rare_truncation <- df_carb_omp %>% select(ompK36_truncations_under_4) %>% rowSums(.) %>% {ifelse(.>0,1,0)}

df_carb_omp <- df_carb_omp %>% select(-muts_under)

# bla genotpyes
bla_count <- df_carb_omp %>% select(colnames(amrfinder_matrix_beta)) %>% colSums() %>% sort
bla_over_4 <- names(bla_count)[bla_count>4] 
bla_under_4 <- names(bla_count)[bla_count<4]

families_under_four <- str_split(bla_under_4,"-",simplify = T) %>% .[,1] %>% gsub("bla","",.)  %>% str_split(.,"_",simplify=T) %>% .[,1]%>% unique %>% sort

store_thes <- lapply(families_under_four,FUN=function(x,df_carb_omp,bla_under_4){
  cols_of_family <- bla_under_4[grepl(x,bla_under_4)]
  new_col_name <- ifelse(x == "ompK35",
                         "Rare_ompK35_insertion",
                         ifelse(x=="ompK36","Rare_ompK36_insertion",paste0("Rare_bla",x))) 
  df_carb_omp[[new_col_name]] <<- df_carb_omp %>% select(all_of(cols_of_family)) %>% rowSums() %>% {ifelse(.>0,1,0)}
},df_carb_omp,bla_under_4)  

df_carb_omp <- df_carb_omp %>% select(-bla_under_4) 
```

# Get matrix 

```{r,fig.width=8.5,height=8.5}
# Carb
df_melt <- df  %>% select(Species_ST,carbapenemase)  %>%
  separate_rows(carbapenemase, sep = " & ") %>%  # split the genotypes
  count(Species_ST,carbapenemase, sort = TRUE)  %>%
    group_by(Species_ST) %>%
    mutate(prop = n / sum(n)) %>%
    ungroup()
 
# Scales 
genotypes <- unique(df$carbapenemase) %>% sort
genotype_colors <- setNames(c(hues::iwanthue(length(genotypes)-1,hmin = 5,cmin = 5,lmin = 5),"gray"), genotypes) 
 
#ST
ST_palette <- ST_scale$palette(1) 
ST_palette <- c(ST_palette,"Other STs"='lightgray')

row_annots <- rowAnnotation(Study = df$study,`Major STs` = df$major_STs,
                            Carbapenemase = df$carbapenemase,
                            col = list(Study = c('2014-15'='blue','2021-23'='red'),Carbapenemase = genotype_colors ,`Major STs`=ST_palette %>% subset(names(.) %in% c(unique(df$major_STs)))), annotation_name_gp = gpar(fontsize = 9),annotation_legend_param = list(
    title_gp = gpar(fontsize = 9),
    labels_gp = gpar(fontsize = 9),
    grid_width = unit(0.25, "cm"),
    grid_height = unit(0.25, "cm")
  ))

porin_bla_genotype <- ComplexHeatmap::Heatmap(df_carb_omp %>%  select(-carbapenemase,-study,-major_STs) %>% mutate_all(as.numeric) %>% select_if(colSums(.,na.rm = T)>0) %>% subset(is.na(rowSums(.))==F) %>% `colnames<-`(recode(colnames(.),"ompK35_rare_truncation"='Rare ompK35 truncation',"ompK36_rare_truncation"='Rare ompK36 truncation',"OmpK36GD" = "GD loop 3 insertion in ompK36","OmpK36TD" = "TD loop 3 insertion in ompK36","OmpK36_c25t" = "C25T mutation in OmpK36") %>% gsub("Rare_","Rare ",.)) ,col=c("white","black"),show_row_names = F,show_row_dend = F,show_column_dend = F,name = "Genotype",gap = unit(0.33, "mm") ,right_annotation = row_annots,
  heatmap_legend_param = list(
    labels = c("1" = "Present","0" = "Absent"),
    labels_gp = gpar(fontsize = 9),
    title_gp = gpar(fontsize =9)
  ),row_names_gp = gpar(fontsize = 9),
    column_names_gp = gpar(fontsize = 9),row_split = df$major_STs,row_title = NULL,cluster_row_slices = T,cluster_column_slices = T
      ) 

porin_bla_genotype

ht_draw <- grid.grabExpr(draw((porin_bla_genotype)))
gghtmp <- cowplot::plot_grid(ht_draw)
```

# Save
```{r}
ggsave(filename = "./figures/s_figure_3_carbapenem_resistance_associated_genotypes_matrix.png",plot=gghtmp,width = 8.5,height = 8.5,dpi=900,bg = 'white')
```
