---
title: "Supplemental Figure 5: Plasmids on phylogeny"
author: "Kyle Gontjes"
date: "`r Sys.Date()`"
output: html_document
---

```{r setup, include=FALSE}
require(knitr)
opts_knit$set(echo = T,warning = FALSE, message = FALSE,error=F,comment=NA) 
opts_chunk$set(echo = F,warning = FALSE, message = FALSE,error=F,comment=NA)
opts_knit$set(root.dir='~/Desktop/gl_nfs/Project_INTEGRATE/Analysis/Resistance_dynamics/INTEGRATE-resistance-ms')
```

# Environment

```{r}
# Packages
packages <- c("tidyverse","ComplexHeatmap",'ggalluvial','ape','ggtree')
lapply(packages,library,character.only=T)

# Functions
source("./lib/consistent_themes.R")

# Print environment
Sys.info()
sessionInfo()
```

# Load data

```{r}
df <- readRDS("./data/dataset/CA_LTACH_first_clinical_CRKP_genomes_merged_studies_metadata.RDS")   
```

# Coordinate ST data

```{r} 
break_of_sts <-  c(sort(unique(df$Species_ST_recode %>% subset(.!="Singleton STs"))),"Singleton STs")
break_of_sts_renamed <- gsub("Klebsiella pneumoniae ","",break_of_sts)
       
ST_scale <- scale_fill_manual(breaks = break_of_sts,values = c(hues::iwanthue(n = length(unique(df$Species_ST_recode)) -1),"lightgray"),labels = break_of_sts_renamed,name = "Sequence type", guide = guide_legend(nrow=3, title.position = "top", label.position = "right"))

df$Species_ST_recode_no_species <- gsub("Klebsiella pneumoniae ","",df$Species_ST)
 
df_201415 <- df%>% subset(study == "2014-15")
df_202123<- df%>% subset(study == "2021-23")
```

# Load carbapenemase data
```{r}
carbapenemase_content <- readRDS("./data/carbapenemase_content/carbapenemase_content_merged_studies.RDS")
df <- left_join(df,carbapenemase_content)
```

# Load plasmid data
```{r} 
KPC_plasmid_matrix <- readRDS("./data/carbapenemase_content/carbapenemase_plasmid_matrix.RDS") %>% subset(rownames(.) %in% df$isolate_no)
```

# Phylogeny 
## Load
```{r}
tree_new <- read.tree("~/Desktop/gl_nfs/Project_INTEGRATE/Sequence_data/variant_calling/2025-04-29_SNPKIT/kpneumoniae/snpkit_rfmt_results/phylo_analysis/IQtree/INTEGRATE_kpneumoniae.treefile")
tree_new_final <- tree_new %>% keep.tip(df$isolate_no %>% subset(. %in% tree_new$tip.label)) %>% phytools::midpoint.root()
```

## Scale
```{r}
feature_colors <- c(`1` = "black",`0`="white")
feature_scale <- scale_fill_manual(breaks = c(1,0), values = c('black','white') ,labels=c("Present","Absent"),name="Presence", guide = guide_legend(nrow=2, title.position = "top", label.position = "right",order=3))
geno_vars <- colnames(KPC_plasmid_matrix %>% select_if(colSums(.)>0)) %>% subset(!.%in% c("study","major_STs","No carbapenemase"))
genotype <- KPC_plasmid_matrix %>% subset(rownames(.) %in% tree_new_final$tip.label) %>% select(geno_vars) %>% drop_na()  %>% select_if(colSums(.) >0) %>% ComplexHeatmap::Heatmap()

genotypes <- unique(df$carbapenemase) %>% sort
genotype_colors <- setNames(c(hues::iwanthue(length(genotypes)-1,hmin = 5,cmin = 5,lmin = 5),"gray"), genotypes)
genotype_scale <- scale_fill_manual(breaks = genotypes,values = genotype_colors,name = "Genotype", guide = guide_legend(ncol=3, title.position = "top", label.position = "right",order=2))   
```

## Phylogeny
```{r,fig.width=10,fig.height=10}  
df <- as.data.frame(df)
rownames(df) <- df$isolate_no
p0 <- gheatmap(ggtree(tree_new_final %>% phytools::midpoint.root()),df %>% select(Species_ST_recode) %>% `colnames<-`("Sequence Type (ST)"),color = NULL,width=0.05,colnames_position = 'top',colnames_angle = 90,hjust=0, colnames_offset_y = 0.25) + ST_scale  + theme(legend.key = element_rect(colour = c('black')))
p01 <- p0 + ggnewscale::new_scale_fill()
p1 <- gheatmap(p01,df %>% select(carbapenemase) %>% `colnames<-`("Carbapenemase"),color = NULL,width=0.05,offset=0.003,colnames_position = 'top',colnames_angle = 90,hjust=0, colnames_offset_y = 0.25)  + genotype_scale + theme( legend.key = element_rect(colour = c('black')))
p11 <- p1 + ggnewscale::new_scale_fill()

figure <- gheatmap(p11,KPC_plasmid_matrix %>% select(genotype@column_names_param$labels[ComplexHeatmap::column_order(genotype)]) %>% mutate_all(as.factor)  %>% `colnames<-`(recode(colnames(.),"ompK35_rare_truncation"='Rare ompK35 truncation',"ompK36_rare_truncation"='Rare ompK36 truncation') %>% trimws(.,"both")),color = NULL,width=0.05*length(geno_vars),offset=0.006,colnames_position = 'top',colnames_angle = 90,hjust=0, colnames_offset_y = 0.25)  + feature_scale + ylim(NA,460) + theme(legend.position = 'bottom', legend.key = element_rect(colour = c('black')))

figure
```

```{r}
ggsave(filename = "./figures/s_figure_5_carbapenem_plasmids_phylogeny.png",plot=figure,width = 10,height =10 ,dpi=900,bg = 'white')
```