---
title: "Phylogeny"
author: "Kyle Gontjes"
date: "`r Sys.Date()`"
output: html_document
---

```{r setup, include=FALSE}
require(knitr)
opts_knit$set(echo = F,warning = FALSE, message = FALSE,error=F,comment=NA) 
opts_chunk$set(echo = F,warning = FALSE, message = FALSE,error=F,comment=NA)
opts_knit$set(root.dir='~/Desktop/gl_nfs/')
```

# Environment

```{r}
# Packages
packages <- c("tidyverse","gridExtra","kableExtra","tableone","gontjesr","cowplot","ComplexHeatmap",'ggalluvial','ape','ggtree')
lapply(packages,library,character.only=T)

# Functions
source("./Project_INTEGRATE/Analysis/Resistance_dynamics/INTEGRATE-resistance-ms/lib/consistent_themes.R")

# Print environment
Sys.info()
sessionInfo()
```

# Load data

```{r}
df <- readRDS("~/Desktop/gl_nfs/Project_INTEGRATE/Analysis/Resistance_dynamics/2025-09-18_master_kpn_first_isolate_dataset_curation/results/CA_clinical_CRKP_first_isolate_metadata_091825.RDS")  
  
df$ST258 <- ifelse(df$Species_ST == "Klebsiella pneumoniae ST258",T,F)
df$ST17 <- ifelse(df$Species_ST == "Klebsiella pneumoniae ST17",T,F)

df$study <- factor(df$study,levels = c("2014-15","2021-23"))
df_201415 <- df%>% subset(study == "2014-15")
df_202123 <- df%>% subset(study == "2021-23")

MIC_variables <- c('MERO',"IMI",'blbli','MVB','IR','CZA',"FDC","CT","OMC","ERV","DLX","CST","PLZ","FOS") 
resistance_vars <- paste0(MIC_variables,"_dich_num")  
```

# Tree
```{r}
tree_new <- read.tree("~/Desktop/gl_nfs/Project_INTEGRATE/Sequence_data/variant_calling/2025-04-29_SNPKIT/kpneumoniae/snpkit_rfmt_results/phylo_analysis/IQtree/INTEGRATE_kpneumoniae.treefile")
tree_new_final <- tree_new %>% keep.tip(df$isolate_no %>% subset(. %in% tree_new$tip.label)) %>% phytools::midpoint.root()
#tree_old <- read.tree("~/Desktop/gl_nfs/Project_Penn_KPC/Sequence_data/variant_calling/2025-02-17_snpkit_Penn_KPC/")

```

```{r}

OLD_ST <- unique(df_201415$Species_ST) %>% sort
new_ST <- unique(df_202123$Species_ST) %>% sort 
ST_over_1 <- table(df$Species_ST) %>% subset(.>1) %>% names

df$Species_ST_new <- ifelse(df$Species_ST %in% c(ST_over_1),df$Species_ST,"Singleton STs") %>% gsub("Klebsiella pneumoniae ","",.)
                                          
break_of_sts <-  c(sort(unique(df$Species_ST_new %>% subset(.!="Singleton STs"))),"Singleton STs")
break_of_sts_renamed <- gsub("Klebsiella pneumoniae ","",break_of_sts)
       
ST_scale <- scale_fill_manual(breaks = break_of_sts,values = c(hues::iwanthue(n = length(unique(df$Species_ST_new)) -1),"lightgray"),labels = break_of_sts_renamed,name = "Sequence type", guide = guide_legend(ncol=7, title.position = "top", label.position = "right"))
```

# Carbapenemase

```{r}
amrfinder <- readRDS("~/Desktop/gl_nfs/Project_INTEGRATE/Analysis/Resistance_dynamics/2025-09-05_master_kpn_dataset_curation/results/amrfinder_cleaned_090525.RDS")
amrfinder <- subset(amrfinder,amrfinder$isolate_no %in% df$isolate_no) 
amrfinder_beta <- amrfinder %>% subset(class == "BETA-LACTAM")

amrfinder_carb_subclass <- subset(amrfinder,subclass == "CARBAPENEM")
amrfinder_carb <- amrfinder %>% subset(subclass == "CARBAPENEM" | grepl("KPC|NDM|VIM|IMP",element_symbol) | grepl("OXA-48 family",element_name))
amrfinder_matrix <- readRDS("~/Desktop/gl_nfs/Project_INTEGRATE/Analysis/Resistance_dynamics/2025-09-05_master_kpn_dataset_curation/results/amrfinder_matrix_090525.RDS")  %>% subset(rownames(.) %in% df$isolate_no)
amrfinder_matrix_beta <- amrfinder_matrix %>% select(any_of(sort(unique(amrfinder_beta$element_symbol))))
amrfinder_matrix_carb <- amrfinder_matrix %>% select(any_of(sort(unique(amrfinder_carb$element_symbol))))

# Melt dataframe matrix so that 
isolate_level_data <- amrfinder_matrix  %>% mutate(isolate_no = rownames(.)) %>% reshape2::melt(id.vars="isolate_no")
isolate_level_data <- left_join(isolate_level_data, df %>% select(isolate_no,Species,Species_ST,study,hospital_location,rectal), by = "isolate_no") 

# Get an isolate's carbapenemase genes
df$blactamase <- sapply(df$isolate_no,FUN=function(x){
  amrfinder_beta %>% subset(isolate_no == x) %>% .$element_symbol %>% paste0(collapse=";")
}) %>% {ifelse(.=="","No blactamase",.)}

df$carbapenemase <- sapply(df$isolate_no,FUN=function(x){
  amrfinder_carb %>% subset(isolate_no == x) %>% .$element_symbol %>% paste0(collapse=";")
}) %>% {ifelse(.=="","No carbapenemase",.)}

df$carbapenemase_family <- sapply(df$isolate_no,FUN=function(x){
    carbapenemase <- amrfinder_carb %>% subset(isolate_no == x) %>% .$element_symbol
     if(is_empty(carbapenemase)){
      "No carbapenemase"
    } else {
      sapply(carbapenemase,FUN=function(y){
        str_split(y,"-",simplify=T) %>% .[,1]
    }) %>% unlist  %>% paste0(collapse=";") 
    }
    
})   
df$blaKPC_2 <- grepl("KPC-2",df$carbapenemase)  & !grepl("KPC-29",df$carbapenemase)
df$no_carbapenemase <- grepl("No carbapenemase",df$carbapenemase_family) 
df_201415 <- df%>% subset(study == "2014-15")
df_202123<- df%>% subset(study == "2021-23")

genotypes <- unique(df$carbapenemase) %>% sort
genotype_colors <- setNames(c(hues::iwanthue(length(genotypes)-1,hmin = 5,cmin = 5,lmin = 5),"gray"), genotypes)

genotype_scale <- scale_fill_manual(breaks = genotypes,values = genotype_colors,name = "Genotype", guide = guide_legend(ncol=4, title.position = "top", label.position = "right"))   
```

# Global kleb 
```{r,fig.width=13,fig.height=15}
MIC_variables <- c('MERO',"IMI",'blbli','MVB','IR','CZA',"FDC","CT","OMC","ERV","DLX","CST","PLZ","FOS") 
resistance_vars <- paste0(MIC_variables,"_dich_num")  
feature_colors <- c(`1` = "black",`0`="white")
feature_scale <- scale_fill_manual(breaks = c(1,0), values = c('black','white') ,labels=c("Present","Absent"),name="Tip state", guide = guide_legend(nrow=1, title.position = "top", label.position = "right"))
resistance_clustering <-df %>% subset(isolate_no %in% tree_new_final$tip.label) %>% select(resistance_vars) %>% drop_na() %>% pheatmap()

df <- as.data.frame(df)
rownames(df) <- df$isolate_no
p0 <- gheatmap(ggtree(tree_new_final %>% phytools::midpoint.root()),df %>% select(Species_ST_new) %>% `colnames<-`("Sequence Type (ST)"),color = NULL,width=0.1,colnames_position = 'top',colnames_angle = 90,hjust=0, colnames_offset_y = 0.25) + ST_scale
p01 <- p0 + ggnewscale::new_scale_fill()
p1 <- gheatmap(p01,df %>% select(carbapenemase) %>% `colnames<-`("Carbapenemase"),color = NULL,width=0.1,offset=0.006,colnames_position = 'top',colnames_angle = 90,hjust=0, colnames_offset_y = 0.25)  + genotype_scale
p11 <- p1 + ggnewscale::new_scale_fill()
A <- gheatmap(p11,df %>% select(resistance_vars[ComplexHeatmap::column_order(resistance_clustering)]) %>% mutate_all(as.factor) %>% `colnames<-`(recode(colnames(.),
                         "CZA_dich_num"='Ceftazidime-avibactam',
                         "CST_dich_num"='Colistin', 
                         'MERO_dich_num'='Meropenem',
                         'IMI_dich_num'='Imipenem',
                         "blbli_dich_num"='BL/BLI agents',
                         "MVB_dich_num"='Meropenem-vaborbactam',
                         "IR_dich_num"='Imipenem-relebactam',
                         "PLZ_dich_num"='Plazomicin',
                         "TMP_SMX_dich_num"='Trimethoprim-Sulfamethoxazole',
                         "OMC_dich_num" = "Omadacycline",
                         "DLX_dich_num" = "Delafloxacin",
                         "ERV_dich_num" = "Eravacycline",
                         "FOS_dich_num" = "Fosfomycin",
                         "CT_dich_num" = "Ceftolozane-tazobactam",
                         "FDC_dich_num"="Cefiderocol")),color = NULL,width=0.1*length(resistance_vars),offset=0.012,colnames_position = 'top',colnames_angle = 90,hjust=0, colnames_offset_y = 0.25)  + feature_scale + ylim(NA,432.5) + theme(legend.position = 'bottom')
A 
```

# ST258
```{r,fig.width=12.5,fig.height=7.5}
st258_df <- df %>% subset(ST=='258')
st258_df <- as.data.frame(st258_df)
rownames(st258_df) <- st258_df$isolate_no
tree_new_final_258 <- tree_new_final %>% keep.tip(subset(st258_df$isolate_no,st258_df$isolate_no %in% tree_new_final$tip.label) )%>% phytools::midpoint.root()
p0 <- gheatmap(ggtree(tree_new_final_258),df %>% select(Species_ST_new)%>% `colnames<-`("Sequence Type (ST)"),color = NULL,width=0.1,colnames_position = 'top',colnames_angle = 90,hjust=0, colnames_offset_y = 0.25) + ST_scale
p01 <- p0 + ggnewscale::new_scale_fill()
p1 <- gheatmap(p01,df %>% select(carbapenemase) %>% `colnames<-`("Carbapenemase"),color = NULL,width=0.1,offset=0.00004,colnames_position = 'top',colnames_angle = 90,hjust=0, colnames_offset_y = 0.25)  + genotype_scale
p11 <- p1 + ggnewscale::new_scale_fill()
B <- gheatmap(p11,df %>% select(resistance_vars[ComplexHeatmap::column_order(resistance_clustering)]) %>% mutate_all(as.factor) %>% `colnames<-`(recode(colnames(.),
                         "CZA_dich_num"='Ceftazidime-avibactam',
                         "CST_dich_num"='Colistin', 
                         'MERO_dich_num'='Meropenem',
                         'IMI_dich_num'='Imipenem',
                         "blbli_dich_num"='BL/BLI agents',
                         "MVB_dich_num"='Meropenem-vaborbactam',
                         "IR_dich_num"='Imipenem-relebactam',
                         "PLZ_dich_num"='Plazomicin',
                         "TMP_SMX_dich_num"='Trimethoprim-Sulfamethoxazole',
                         "OMC_dich_num" = "Omadacycline",
                         "DLX_dich_num" = "Delafloxacin",
                         "ERV_dich_num" = "Eravacycline",
                         "FOS_dich_num" = "Fosfomycin",
                         "CT_dich_num" = "Ceftolozane-tazobactam",
                         "FDC_dich_num"="Cefiderocol")),color = NULL,width=0.1*length(resistance_vars),offset=0.00008,colnames_position = 'top',colnames_angle = 90,hjust=0, colnames_offset_y = 0.25)  + feature_scale + ylim(NA,222.5) + ggtitle("ST258") + theme(legend.position = 'none')
B
```


# 14
```{r,fig.width=9,fig.height=7.5}
st14_df <- df %>% subset(ST=='14')
st14_df <- as.data.frame(st14_df)
rownames(st14_df) <- st14_df$isolate_no
tree_new_final_14 <- tree_new_final %>% keep.tip(subset(st14_df$isolate_no,st14_df$isolate_no %in% tree_new_final$tip.label) )%>% phytools::midpoint.root()
p0 <- gheatmap(ggtree(tree_new_final_14),df %>% select(Species_ST_new)%>% `colnames<-`("Sequence Type (ST)"),color = NULL,width=0.1,colnames_position = 'top',colnames_angle = 90,hjust=0, colnames_offset_y = 0.25) + ST_scale
p01 <- p0 + ggnewscale::new_scale_fill()
p1 <- gheatmap(p01,df %>% select(carbapenemase) %>% `colnames<-`("Carbapenemase"),color = NULL,width=0.1,offset=0.000055,colnames_position = 'top',colnames_angle = 90,hjust=0, colnames_offset_y = 0.25)  + genotype_scale
p11 <- p1 + ggnewscale::new_scale_fill()
C <- gheatmap(p11,df %>% select(resistance_vars[ComplexHeatmap::column_order(resistance_clustering)]) %>% mutate_all(as.factor) %>% `colnames<-`(recode(colnames(.),
                         "CZA_dich_num"='Ceftazidime-avibactam',
                         "CST_dich_num"='Colistin', 
                         'MERO_dich_num'='Meropenem',
                         'IMI_dich_num'='Imipenem',
                         "blbli_dich_num"='BL/BLI agents',
                         "MVB_dich_num"='Meropenem-vaborbactam',
                         "IR_dich_num"='Imipenem-relebactam',
                         "PLZ_dich_num"='Plazomicin',
                         "TMP_SMX_dich_num"='Trimethoprim-Sulfamethoxazole',
                         "OMC_dich_num" = "Omadacycline",
                         "DLX_dich_num" = "Delafloxacin",
                         "ERV_dich_num" = "Eravacycline",
                         "FOS_dich_num" = "Fosfomycin",
                         "CT_dich_num" = "Ceftolozane-tazobactam",
                         "FDC_dich_num"="Cefiderocol")),color = NULL,width=0.1*length(resistance_vars),offset=0.00011,colnames_position = 'top',colnames_angle = 90,hjust=0, colnames_offset_y = 0.25)  + feature_scale + ylim(NA,128.5) + ggtitle("ST14") + theme(legend.position = 'none')
C
```

# ST17
```{r,fig.width=9,fig.height=7.5}
st17_df <- df %>% subset(ST=='17')
st17_df <- as.data.frame(st17_df)
rownames(st17_df) <- st17_df$isolate_no
tree_new_final_17 <- tree_new_final %>% keep.tip(subset(st17_df$isolate_no,st17_df$isolate_no %in% tree_new_final$tip.label) )%>% phytools::midpoint.root()
p0 <- gheatmap(ggtree(tree_new_final_17),df %>% select(Species_ST_new)%>% `colnames<-`("Sequence Type (ST)"),color = NULL,width=0.1,colnames_position = 'top',colnames_angle = 90,hjust=0, colnames_offset_y = 0.25) + ST_scale
p01 <- p0 + ggnewscale::new_scale_fill()
p1 <- gheatmap(p01,df %>% select(carbapenemase) %>% `colnames<-`("Carbapenemase"),color = NULL,width=0.1,offset=0.000015,colnames_position = 'top',colnames_angle = 90,hjust=0, colnames_offset_y = 0.25)  + genotype_scale
p11 <- p1 + ggnewscale::new_scale_fill()
D <- gheatmap(p11,df %>% select(resistance_vars[ComplexHeatmap::column_order(resistance_clustering)]) %>% mutate_all(as.factor) %>% `colnames<-`(recode(colnames(.),
                         "CZA_dich_num"='Ceftazidime-avibactam',
                         "CST_dich_num"='Colistin', 
                         'MERO_dich_num'='Meropenem',
                         'IMI_dich_num'='Imipenem',
                         "blbli_dich_num"='BL/BLI agents',
                         "MVB_dich_num"='Meropenem-vaborbactam',
                         "IR_dich_num"='Imipenem-relebactam',
                         "PLZ_dich_num"='Plazomicin',
                         "TMP_SMX_dich_num"='Trimethoprim-Sulfamethoxazole',
                         "OMC_dich_num" = "Omadacycline",
                         "DLX_dich_num" = "Delafloxacin",
                         "ERV_dich_num" = "Eravacycline",
                         "FOS_dich_num" = "Fosfomycin",
                         "CT_dich_num" = "Ceftolozane-tazobactam",
                         "FDC_dich_num"="Cefiderocol")),color = NULL,width=0.1*length(resistance_vars),offset=0.00003,colnames_position = 'top',colnames_angle = 90,hjust=0, colnames_offset_y = 0.25)  + feature_scale + ylim(NA,42.5)+ ggtitle("ST17") + theme(legend.position = 'none')
D
```

# ST307
```{r,fig.width=9,fig.height=7.5}
st307_df <- df %>% subset(ST=='307')
st307_df <- as.data.frame(st307_df)
rownames(st307_df) <- st307_df$isolate_no
tree_new_final_307 <- tree_new_final %>% keep.tip(subset(st307_df$isolate_no,st307_df$isolate_no %in% tree_new_final$tip.label) )%>% phytools::midpoint.root()
p0 <- gheatmap(ggtree(tree_new_final_307),df %>% select(Species_ST_new)%>% `colnames<-`("Sequence Type (ST)"),color = NULL,width=0.1,colnames_position = 'top',colnames_angle = 90,hjust=0, colnames_offset_y = 0.25) + ST_scale
p01 <- p0 + ggnewscale::new_scale_fill()
p1 <- gheatmap(p01,df %>% select(carbapenemase) %>% `colnames<-`("Carbapenemase"),color = NULL,width=0.1,offset=0.00002,colnames_position = 'top',colnames_angle = 90,hjust=0, colnames_offset_y = 0.25)  + genotype_scale
p11 <- p1 + ggnewscale::new_scale_fill()
E <- gheatmap(p11,df %>% select(resistance_vars[ComplexHeatmap::column_order(resistance_clustering)]) %>% mutate_all(as.factor) %>% `colnames<-`(recode(colnames(.),
                         "CZA_dich_num"='Ceftazidime-avibactam',
                         "CST_dich_num"='Colistin', 
                         'MERO_dich_num'='Meropenem',
                         'IMI_dich_num'='Imipenem',
                         "blbli_dich_num"='BL/BLI agents',
                         "MVB_dich_num"='Meropenem-vaborbactam',
                         "IR_dich_num"='Imipenem-relebactam',
                         "PLZ_dich_num"='Plazomicin',
                         "TMP_SMX_dich_num"='Trimethoprim-Sulfamethoxazole',
                         "OMC_dich_num" = "Omadacycline",
                         "DLX_dich_num" = "Delafloxacin",
                         "ERV_dich_num" = "Eravacycline",
                         "FOS_dich_num" = "Fosfomycin",
                         "CT_dich_num" = "Ceftolozane-tazobactam",
                         "FDC_dich_num"="Cefiderocol")),color = NULL,width=0.1*length(resistance_vars),offset=0.00004,colnames_position = 'top',colnames_angle = 90,hjust=0, colnames_offset_y = 0.25)  + feature_scale + ylim(NA,42.5)+ ggtitle("ST307") + theme(legend.position = 'none') 
E
```

# Combine figures
```{r,fig.width=13,fig.height=16}
A_fig <- cowplot::plot_grid(A + theme(legend.position='none'),labels = c("A"))
A_legend <- cowplot::get_legend(A)
legend_fig <- cowplot::plot_grid(A_legend)
smaller <- cowplot::plot_grid(B,C,D,E,ncol=2,labels = c("B","C","D","E"),hjust = 0.25)
merged <- cowplot::plot_grid(A_fig,smaller,ncol=2,rel_widths = c(0.5,1))
figure <- cowplot::plot_grid(merged,A_legend,rel_heights = c(1,0.075),ncol=1)
figure
```



```{r,fig.width=17,fig.height=10.5}   
ggsave(filename = "./Project_INTEGRATE/Analysis/Resistance_dynamics/INTEGRATE-resistance-ms/figures/s_figure_X_resistance_phylogeny.png",plot=figure,width = 13,height = 16,dpi=900,bg = 'white')
```