---
title: "Supplemental figure 3: Genotype on phylogeny"
author: "Kyle Gontjes"
date: "`r Sys.Date()`"
output: html_document
---

```{r setup, include=FALSE}
require(knitr)
opts_knit$set(echo = F,warning = FALSE, message = FALSE,error=F,comment=NA) 
opts_chunk$set(echo = F,warning = FALSE, message = FALSE,error=F,comment=NA)
opts_knit$set(root.dir='~/Desktop/gl_nfs/Project_INTEGRATE/Analysis/Resistance_dynamics/INTEGRATE-resistance-ms')
```

```{r}
# Packages
packages <- c("tidyverse","gridExtra","kableExtra","tableone","gontjesr","cowplot","ComplexHeatmap",'ggalluvial','ape','ggtree')
lapply(packages,library,character.only=T)

# Functions
source("./lib/consistent_themes.R")

# Print environment
Sys.info()
sessionInfo()
```

# Load data

```{r}
df <- readRDS("./data/dataset/CA_LTACH_first_clinical_per_quarter_CRKP_genomes_merged_studies_metadata.RDS")   
   
df_201415 <- df%>% subset(study == "2014-15")
df_202123 <- df%>% subset(study == "2021-23") 
```

# Coordinate ST data

```{r} 
break_of_sts <-  c(sort(unique(df$Species_ST_recode %>% subset(.!="Singleton STs"))),"Singleton STs")
break_of_sts_renamed <- gsub("Klebsiella pneumoniae ","",break_of_sts)
       
ST_scale <- scale_fill_manual(breaks = break_of_sts,values = c(hues::iwanthue(n = length(unique(df$Species_ST_recode)) -1),"lightgray"),labels = break_of_sts_renamed,name = "Sequence type", guide = guide_legend(nrow=3, title.position = "top", label.position = "right",order=1))

ST_palette <- ST_scale$palette(1) 
ST_palette <- c(ST_palette,"Other STs"='lightgray')
ST_major_scale <- scale_fill_manual(values = ST_palette,name = "Major sequence types (STs)")
```

# Load  carbapenemase content
```{r}
carbapenemase_content <- readRDS("./data/carbapenemase_content/carbapenemase_content_merged_studies.RDS")
df <- left_join(df,carbapenemase_content)
```

# Load omp matrix
```{r}
carb_omp <-  readRDS("./data/carbapenemase_content/carbapenem_resistance_genotype_content_merged_first_clinical_CRKP_isolate_per_quarter.RDS")
``` 

# Supplemental Figure:
```{r,fig.width=9.5,height=6.7857}  
df_carb_omp <- df %>% select(isolate_no,carbapenemase,study,major_STs) %>% left_join(., carb_omp %>% mutate(isolate_no = rownames(.)))    %>% as.data.frame
rownames(df_carb_omp) <- df_carb_omp$isolate_no 
```

# Phylogeny 
```{r}
tree_new <- read.tree("~/Desktop/gl_nfs/Project_INTEGRATE/Sequence_data/variant_calling/2025-04-29_SNPKIT/kpneumoniae/snpkit_rfmt_results/phylo_analysis/IQtree/INTEGRATE_kpneumoniae.treefile")
tree_new_final <- tree_new %>% keep.tip(df$isolate_no %>% subset(. %in% tree_new$tip.label)) %>% phytools::midpoint.root()
```

```{r,fig.width=11,fig.height=10}  
feature_colors <- c(`1` = "black",`0`="white")
feature_scale <- scale_fill_manual(breaks = c(1,0), values = c('black','white') ,labels=c("Present","Absent"),name="Genotype", guide = guide_legend(nrow=3,order=3, title.position = "top", label.position = "right"))
geno_vars <- colnames(df_carb_omp) %>% subset(!.%in% c("study","major_STs","isolate_no"))
genotype <- df_carb_omp %>% subset(rownames(.) %in% tree_new_final$tip.label) %>% select(geno_vars) %>% select(-carbapenemase) %>% drop_na() %>% select_if(colSums(.)>0) %>% ComplexHeatmap::Heatmap()

genotypes <- c(table(df$carbapenemase %>% subset(.!="No carbapenemase")) %>% sort(decreasing = T) %>% names,"No carbapenemase")
genotype_colors <- setNames(c(hues::iwanthue(length(genotypes)-1,hmin = 5,cmin = 5,lmin = 5),"gray"), genotypes)
genotype_scale <- scale_fill_manual(breaks = genotypes,values = genotype_colors,name = "Carbapenemase", guide = guide_legend(nrow=3, title.position = "top", label.position = "right",order=2)) 

df <- as.data.frame(df)
rownames(df) <- df$isolate_no
p0 <- gheatmap(ggtree(tree_new_final %>% phytools::midpoint.root()),df %>% select(Species_ST_recode) %>% `colnames<-`("Sequence Type (ST)"),color = NULL,width=0.05,colnames_position = 'top',colnames_angle = 90,hjust=0, colnames_offset_y = 0.25) + ST_scale +  theme(legend.key = element_rect(colour = c('black')))
p01 <- p0 + ggnewscale::new_scale_fill()
p1 <- gheatmap(p01,df %>% select(carbapenemase) %>% `colnames<-`("Carbapenemase"),color = NULL,width=0.05,offset=0.0031,colnames_position = 'top',colnames_angle = 90,hjust=0, colnames_offset_y = 0.25)  + genotype_scale + theme(legend.key = element_rect(colour = c('black')))
p11 <- p1 + ggnewscale::new_scale_fill()

figure <- gheatmap(p11,df_carb_omp %>% select(genotype@column_names_param$labels[ComplexHeatmap::column_order(genotype)]) %>% mutate_all(as.factor)  %>% `colnames<-`(recode(colnames(.),"ompK35_rare_truncation"='Rare ompK35 truncation',"ompK36_rare_truncation"='Rare ompK36 truncation',"OmpK36GD" = "GD loop 3 insertion in ompK36","OmpK36TD" = "TD loop 3 insertion in ompK36","OmpK36_c25t" = "C25T mutation in OmpK36") %>% gsub("Rare_","Rare ",.)),color  = NULL,width=0.05*length(geno_vars),offset=0.006,colnames_position = 'top',colnames_angle = 90,hjust=0, colnames_offset_y = 0.25)  + feature_scale + ylim(NA,610) + theme(legend.position = 'bottom',legend.direction="horizontal", 
                                    legend.justification = "center", legend.key = element_rect(colour = c('black')))  + geom_treescale(y = -10)
figure
```

```{r}
ggsave(filename = "./figures/s_figure_3_carbapenem_resistance_associated_genotypes_phylogeny.png",plot=figure, width = 11,height = 10,dpi=900,bg = 'white')
``` 
