---
title: "Supplemental Figure 6: Plasmid data"
author: "Kyle Gontjes"
date: "`r Sys.Date()`"
output: html_document
---

```{r setup, include=FALSE}
require(knitr)
opts_knit$set(echo = T,warning = FALSE, message = FALSE,error=F,comment=NA) 
opts_chunk$set(echo = T,warning = FALSE, message = FALSE,error=F,comment=NA)
opts_knit$set(root.dir='~/Desktop/gl_nfs/Project_INTEGRATE/Analysis/Resistance_dynamics/INTEGRATE-resistance-ms')
```

# Environment

```{r}
# Packages
packages <- c("tidyverse","gridExtra","kableExtra","tableone","gontjesr","cowplot","ComplexHeatmap",'ggalluvial')
lapply(packages,library,character.only=T)

# Functions
source("./lib/consistent_themes.R")

# Print environment
Sys.info()
sessionInfo()
```

# Load data
```{r} 
df <- readRDS("./data/dataset/CA_LTACH_first_clinical_per_quarter_CRKP_genomes_merged_studies_metadata.RDS")    
```


# Load carbapenemase data
```{r}
carbapenemase_content <- readRDS("./data/carbapenemase_content/carbapenemase_content_merged_studies.RDS")
df <- left_join(df,carbapenemase_content)
```

# Load plasmid data

```{r} 
mobtyper_merged <- readRDS("./data/mobsuite/mobtyper_merged_w_carbapenemase_data.RDS") %>% subset(isolate_no %in% df$isolate_no) 
most_frequent_plasmids <- names(mobtyper_merged$primary_cluster_id  %>% table %>% sort(decreasing = T))

mobtyper_merged$primary_cluster_id <- factor(mobtyper_merged$primary_cluster_id,levels =most_frequent_plasmids)
mobsuite_contigs <-  readRDS("./data/mobsuite/mobsuite_contig_merged_w_carbapenemase_data.RDS") %>% subset(isolate_no %in% df$isolate_no)  
mobsuite_contigs$primary_cluster_id <- factor(mobsuite_contigs$primary_cluster_id,levels =most_frequent_plasmids)
```

# Study distribution

```{r}
mobtyper_sum <- mobtyper_merged %>% group_by(primary_cluster_id) %>% summarise(n=n(),carbapenemase_plasmid = sum(carbapenemase_plasmid),Non_carbapenemase_plasmid = n-carbapenemase_plasmid)  
KPC_plasmids <- subset(mobtyper_sum,carbapenemase_plasmid>0) %>% arrange(-carbapenemase_plasmid) %>%  .$primary_cluster_id
 
KPC_plasmid_scale <-  scale_fill_manual(breaks = c(1,0),values=c("red","gray"),labels = c("Present","Absent"),name = "Carbapenemase content",guide = guide_legend(nrow=2, title.position = "top", label.position = "right"))
KPC_contig_bool <-  scale_fill_manual(breaks = c(T,F),values=c("red","gray"),labels = c("Present","Absent"),name = "Carbapenemase content",guide = guide_legend(nrow=2, title.position = "top", label.position = "right"))

mobtyper_merged$carbapenemase_plasmid <- as.factor(mobtyper_merged$carbapenemase_plasmid) 

A <- ggplot(data=mobtyper_merged %>% subset(primary_cluster_id %in% KPC_plasmids),aes(x=primary_cluster_id,fill=carbapenemase_plasmid)) + geom_bar(stat='count')  + KPC_plasmid_scale  + 
  theme_bw_me + theme(axis.text.x = element_text(angle=45,hjust=1)) + 
  labs(x="Primary MOB-cluster",y="Isolates",fill="blaKPC-containing plasmid") + facet_grid(~study)
A
```

# Carbapenemase data

```{r}
genotypes <- c(table(df$carbapenemase %>% subset(.!="No carbapenemase")) %>% sort(decreasing = T) %>% names,"No carbapenemase")
genotype_colors <- setNames(c(hues::iwanthue(length(genotypes)-1,hmin = 5,cmin = 5,lmin = 5),"gray"), genotypes)

genotype_scale <- scale_fill_manual(breaks = genotypes,values = genotype_colors,name = "Carbapenemase", guide = guide_legend(nrow=2, title.position = "top", label.position = "right")) 


B <- ggplot(data=mobtyper_merged %>% subset(primary_cluster_id %in% KPC_plasmids) %>% subset(carbapenemase_plasmid==1),aes(x=primary_cluster_id,fill=carbapenemase)) + geom_bar(stat='count')  + genotype_scale + 
  theme_bw_me + theme(axis.text.x = element_text(angle=45,hjust=1)) + 
  labs(x="Primary MOB-cluster",y="Isolates",fill="Carbapenemase")   + facet_grid(~study)  

B
```

# C. Contig count

```{r}
C <-  ggplot(data=mobtyper_merged %>% subset(primary_cluster_id %in% KPC_plasmids),aes(x=primary_cluster_id,fill=carbapenemase_plasmid),y=num_contigs) + geom_boxplot(aes(y=num_contigs)) + KPC_plasmid_scale + 
  theme_bw_me + theme(axis.text.x = element_text(angle=45,hjust=1)) + 
  labs(x="",y="Number of contigs",fill="Carbapenemase-containing plasmid")  

C
```

# D. Size

```{r}
D <-  ggplot(data=mobtyper_merged %>% subset(primary_cluster_id %in% KPC_plasmids),aes(x=primary_cluster_id,fill=carbapenemase_plasmid),y=size) + geom_boxplot(aes(y=size))  + KPC_plasmid_scale + 
  theme_bw_me + theme(axis.text.x = element_text(angle=45,hjust=1)) + 
  labs(x="",y="Size of plasmid (bp)",fill="blaKPC-containing plasmid")  + scale_y_continuous(breaks = c(0,10000,25000,50000,75000,100000,150000,200000,250000,300000,350000))

D
```

# E. KPC contig size 

```{r} 
KPC_contig_data <- mobsuite_contigs %>% subset(primary_cluster_id != "-" & primary_cluster_id %in% KPC_plasmids) 
E <-  ggplot(data=KPC_contig_data  ,aes(x=primary_cluster_id,fill=carbapenemase_contig),y=size) + geom_boxplot(aes(y=size))  + KPC_contig_bool + 
  theme_bw_me + theme(axis.text.x = element_text(angle=45,hjust=1)) + 
  labs(x="",y="Contig size (bp)",fill="carbapenemase-containing contig")  + geom_hline(yintercept = 10000,linetype = 2)  + scale_y_continuous(breaks = c(0,10000,25000,50000,75000,100000,150000,200000))

E
```

# F. Predicted Mobility

```{r}
mobility_scale <- scale_fill_manual(values = c("brown","blue","#C0C0C0"),breaks = c("conjugative",'mobilizable','non-mobilizable'),labels = c("Conjugative","Mobilizable","Non-Mobilizable"),name="Predicted mobility",guide = guide_legend(nrow=2, title.position = "top", label.position = "right"))

## Predicted mobility 
mobtyper_merged$KPC_plasmid_name <- ifelse(mobtyper_merged$carbapenemase_plasmid ==1,"Carbapenemase plasmid","Non-carbapenemase plasmid")
f <- ggplot(data=mobtyper_merged%>% subset(primary_cluster_id %in% KPC_plasmids),aes(x=primary_cluster_id,fill=predicted_mobility))  + geom_bar()+ facet_wrap(~KPC_plasmid_name)  +   labs(x="",y="Isolates",fill="Predicted mobility") + theme_bw_me + 
  theme_bw_me + theme(axis.text.x = element_text(angle=45,hjust=1),
                      strip.background = element_rect(color = 'black',linewidth = 1)) + mobility_scale
  
f
```

# G. Observed host range

```{r}
color_range <- hues::iwanthue(n = 6,hmin = 0,hmax = 75,plot = T)

host_range_scale <- scale_fill_manual(values = c("darkgray", color_range ),breaks = c("Actinomycetota,Bacillota,Pseudomonadota","Klebsiella","Enterobacteriaceae","Enterobacterales"                            ,"Gammaproteobacteria","Salmonella"),labels =  c("Actinomycetota, Bacillota, Pseudomonadota","Klebsiella","Enterobacteriaceae","Enterobacterales"                            ,"Gammaproteobacteria","Salmonella"),name="Observed host range",guide = guide_legend(nrow=2, title.position = "top", label.position = "right"))

G <-    ggplot(data=mobtyper_merged%>% subset(primary_cluster_id %in% KPC_plasmids),aes(x=primary_cluster_id,fill=observed_host_range_ncbi_name))  + geom_bar()+ facet_wrap(~KPC_plasmid_name)    +  labs(x="",y="Isolates",fill="Observed host range")+ theme_bw_me + 
  theme_bw_me + theme(axis.text.x = element_text(angle=45,hjust=1),
                      strip.background = element_rect(color = 'black',linewidth = 1)) + host_range_scale

G
```

# H. Replicon type

```{r} 
mobtyper_merged$rep_types <- mobtyper_merged$`rep_type(s)`
rep_type_matrix <- mobtyper_merged %>% subset(primary_cluster_id %in% KPC_plasmids) %>% select(isolate_no,primary_cluster_id,rep_types)%>%
  separate_rows(rep_types,sep=",") %>%
  group_by(primary_cluster_id,rep_types) %>%
  summarise(n=n(),.groups='drop') %>% as.data.frame %>% mutate(rep_types = recode(rep_types,"-" = 'Absent')) %>% 
  pivot_wider(names_from = rep_types,values_from = n,values_fill = 0) %>%
  as.data.frame() %>% `rownames<-`(.$primary_cluster_id) %>% select(-primary_cluster_id)  

H <- ComplexHeatmap::Heatmap(rep_type_matrix,col=c('white','red'),name = 'Count',cell_fun =  function(j, i, x, y, width, height, fill) {
  if(rep_type_matrix[i, j]>0){
    grid.text(sprintf("%.3s", rep_type_matrix[i, j]), x, y, gp = gpar(fontsize = 11))
  }
},show_heatmap_legend = F,cluster_rows = F,column_title = "Replicon type",height = unit(8, "cm"),show_column_dend = T,width =ncol(rep_type_matrix)*unit(0.75, "cm"), column_names_max_height = unit(6, "cm"))

H

maxchar <- colnames(rep_type_matrix) %>% sapply(nchar) %>% max
```

# I. Relaxase type

```{r}
mobtyper_merged$relaxase_types <- mobtyper_merged$`relaxase_type(s)`

relaxase_type_matrix <- mobtyper_merged  %>% subset(primary_cluster_id %in% KPC_plasmids) %>% select(isolate_no,primary_cluster_id,relaxase_types)%>%
  separate_rows(relaxase_types,sep=",") %>%
  group_by(primary_cluster_id,relaxase_types) %>%
  summarise(n=n(),.groups='drop')  %>% as.data.frame %>% mutate(relaxase_types = recode(relaxase_types,"-" = 'Absent')) %>%
  pivot_wider(names_from = relaxase_types,values_from = n,values_fill = 0) %>%
  as.data.frame() %>% `rownames<-`(.$primary_cluster_id) %>% select(-primary_cluster_id) 

colnames(relaxase_type_matrix) <- sapply(colnames(relaxase_type_matrix),FUN=function(x){
  charl <- nchar(x)
  padding <- maxchar - charl +10
  paste0(paste0(rep(" ",padding),collapse=""),x,collapse = "")
})

I <- ComplexHeatmap::Heatmap(relaxase_type_matrix,col=c('white','red'),name = 'Count',cell_fun =  function(j, i, x, y, width, height, fill) {
  if(relaxase_type_matrix[i, j]>0){
    grid.text(sprintf("%.3s", relaxase_type_matrix[i, j]), x, y, gp = gpar(fontsize = 11))
  }
},show_heatmap_legend = F,cluster_rows = F,column_title = "Relaxase type",height = unit(8, "cm"),show_column_dend = T, width =ncol(relaxase_type_matrix)*unit(0.75, "cm"), column_names_max_height = unit(5, "cm"))

I
```

# J. MPF Type

```{r} 
mpf_type_matrix <- mobtyper_merged  %>% subset(primary_cluster_id %in% KPC_plasmids)  %>% select(isolate_no,primary_cluster_id,mpf_type)%>%
  separate_rows(mpf_type,sep=",") %>%
  group_by(primary_cluster_id,mpf_type) %>%
  summarise(n=n(),.groups='drop') %>% as.data.frame %>% mutate(mpf_type = recode(mpf_type,"-" = 'Absent'))  %>%
  pivot_wider(names_from = mpf_type,values_from = n,values_fill = 0) %>%
  as.data.frame() %>% `rownames<-`(.$primary_cluster_id) %>% select(-primary_cluster_id) 

colnames(mpf_type_matrix) <- sapply(colnames(mpf_type_matrix),FUN=function(x){
  charl <- nchar(x)
  padding <- maxchar - charl +10
  paste0(paste0(rep(" ",padding),collapse=""),x,collapse = "")
})

J <- ComplexHeatmap::Heatmap(mpf_type_matrix,col=c('white','red'),name = 'Count',cell_fun =  function(j, i, x, y, width, height, fill) {
  if(mpf_type_matrix[i, j]>0){
    grid.text(sprintf("%.3s", mpf_type_matrix[i, j]), x, y, gp = gpar(fontsize = 11))
  }
},show_heatmap_legend = F,cluster_rows = F,column_title = "MPF type",height = unit(8, "cm"),show_column_dend = T,width =ncol(mpf_type_matrix)*unit(0.75, "cm"), column_names_max_height = unit(5, "cm"))

J
```

# K. oriT type

```{r}
mobtyper_merged$orit_types <- mobtyper_merged$`orit_type(s)`
orit_type_matrix <- mobtyper_merged %>% subset(primary_cluster_id %in% KPC_plasmids) %>% select(isolate_no,primary_cluster_id,orit_types)%>%
  separate_rows(orit_types,sep=",") %>%
  group_by(primary_cluster_id,orit_types) %>%
  summarise(n=n(),.groups='drop') %>% as.data.frame %>% mutate(orit_types = recode(orit_types,"-" = 'Absent')) %>%
  pivot_wider(names_from = orit_types,values_from = n,values_fill = 0) %>%
  as.data.frame() %>% `rownames<-`(.$primary_cluster_id) %>% select(-primary_cluster_id) 

colnames(orit_type_matrix) <- sapply(colnames(orit_type_matrix),FUN=function(x){
  charl <- nchar(x)
 padding <- maxchar - charl+10
  paste0(paste0(rep(" ",padding),collapse=""),x,collapse = "")
})

K <- ComplexHeatmap::Heatmap(orit_type_matrix,col=c('white','red'),name = 'Count',cell_fun =  function(j, i, x, y, width, height, fill) {
  if(orit_type_matrix[i, j]>0){
    grid.text(sprintf("%.3s", orit_type_matrix[i, j]), x, y, gp = gpar(fontsize = 11))
  }
},show_heatmap_legend = F,cluster_rows = F,column_title = "oriT type",height = unit(8, "cm"),width = ncol(orit_type_matrix)*unit(0.75, "cm"),show_column_dend = T, column_names_max_height = unit(5, "cm"))

K 
```

# Grouped figure

```{r,fig.height=18.75,fig.width=15}
fig_histograms <- cowplot::plot_grid(A,B ,NULL,ncol=3,labels = c("A","B",""),rel_widths = c(1,1,0.2))
fig_histograms2 <- cowplot::plot_grid(C +theme(legend.position='none'),D+ theme(legend.position='none'),E + theme(legend.position='none'),ncol=3,labels = c("C","D","E"))

fig_facet_histograms <-  cowplot::plot_grid(f,G,NULL,ncol=3,rel_widths=c(1,1,0.2),labels = c("F","G","")) 
  
I_mat <-  grid.grabExpr(draw(H))
J_mat <-  grid.grabExpr(draw(I))
K_mat <-  grid.grabExpr(draw(J))
L_mat <-  grid.grabExpr(draw(K))

heatmaps_cow <- cowplot::plot_grid(I_mat,J_mat,K_mat,L_mat,nrow=1,labels = c('H','I',"J",'K'),rel_widths = c(1,0.4,0.4,0.4),align='hv')

figure <- cowplot::plot_grid(fig_histograms,fig_histograms2,fig_facet_histograms,heatmaps_cow,ncol=1,rel_heights=c(0.66,0.5,0.66,1))

figure
```

# Save figure
```{r}
ggsave("./figures/s_figure_6_mobtyper_plasmid_overview.png",plot=figure, units='in',height=18.75,width=15,dpi=900,bg='white')
```

