---
title: "Generate good outgroup for INTEGRATE"
author: "Kyle Gontjes"
date: "`r Sys.Date()`"
output: html_document
---

```{r setup, include=FALSE}
require(knitr)
opts_knit$set(echo = F,warning = FALSE, message = FALSE,error=F,comment=NA) 
opts_chunk$set(echo = F,warning = FALSE, message = FALSE,error=F,comment=NA)
opts_knit$set(root.dir='~/Desktop/gl_nfs/Project_INTEGRATE/Analysis/Resistance_dynamics/INTEGRATE-resistance-ms')
```

# Environment

```{r}
# Packages
packages <- c("tidyverse","gridExtra","kableExtra","tableone","gontjesr","cowplot","ComplexHeatmap",'ggalluvial','ape','ggtree','phytools')
lapply(packages,library,character.only=T)

# Functions
source("./lib/consistent_themes.R")

# Print environment
Sys.info()
sessionInfo()
```

# Load data
```{r}
fasta <- read.FASTA("./data/phylogeny/INTEGRATE/INTEGRATE_cognac_SNP_sites.fa",type = 'DNA')
dist_mat <- dist.dna(fasta,model="N",pairwise.deletion = T,as.matrix = T)

df <- readRDS("./data/dataset/CA_LTACH_first_clinical_per_quarter_CRKP_genomes_merged_studies_metadata.RDS") %>% subset(study == "2021-23")  
df_ST258 <- df %>% subset(ST=="ST258")
df_nonST258 <- df %>% subset(ST!='ST258')
```

# Get summary stats
1. Advice: Choose the ST163 genomes as outgroup - they are distant enough to root the tree but far enough to provide good resolution within ST258 clade
```{r}
candidate_outgroups <- df_nonST258$isolate_no
comp_group = df_ST258$isolate_no

get_summary_stats_distance <- function(isolate,comp_group,dist_mat){
  comp_list <- dist_mat[isolate,comp_group] 
  isolate_ST <- subset(df,isolate_no == isolate) %>% .$ST
  min_val <- min(comp_list)
  median_val <- median(comp_list)
  IQR_val <- IQR(comp_list)
  max_val <- max(comp_list)
  cbind.data.frame(isolate,isolate_ST,min_val,median_val,IQR_val,max_val)
}

outgroup_candidates <- lapply(candidate_outgroups,get_summary_stats_distance,comp_group,dist_mat) %>% do.call(rbind,.) 
outgroup_candidates <- outgroup_candidates %>% arrange(-median_val)

head(outgroup_candidates)

ggplot(data=outgroup_candidates,aes(x=median_val)) + geom_histogram()+ theme_bw() + xlab("Median distance to ST258 genomes") + ylab("Count of candidate outgroup genomes")
```

