---
title: "Results section 4 & 5: Carbapenemase and plasmids"
author: "Kyle Gontjes"
date: "`r Sys.Date()`"
output: html_document
---

```{r setup, include=FALSE}
require(knitr)
opts_knit$set(echo = F,warning = FALSE, message = FALSE,error=F,comment=NA) 
opts_chunk$set(echo = F,warning = FALSE, message = FALSE,error=F,comment=NA)
opts_knit$set(root.dir='~/Desktop/gl_nfs/')
```

# Environment

```{r}
# Packages
packages <- c("tidyverse","gridExtra","kableExtra","tableone","gontjesr","cowplot","ComplexHeatmap",'ggalluvial')
lapply(packages,library,character.only=T)

# Functions
source("./Project_INTEGRATE/Analysis/Resistance_dynamics/INTEGRATE-resistance-ms/lib/consistent_themes.R")

# Print environment
Sys.info()
sessionInfo()
```

# Load data

```{r}
df <- readRDS("~/Desktop/gl_nfs/Project_INTEGRATE/Analysis/Resistance_dynamics/2025-09-18_master_kpn_first_isolate_dataset_curation/results/CA_clinical_CRKP_first_isolate_metadata_091825.RDS")  
  
df$ST258 <- ifelse(df$Species_ST == "Klebsiella pneumoniae ST258",T,F)
df$ST17 <- ifelse(df$Species_ST == "Klebsiella pneumoniae ST17",T,F)

df$study <- factor(df$study,levels = c("2014-15","2021-23"))
df_201415 <- df%>% subset(study == "2014-15")
df_202123 <- df%>% subset(study == "2021-23")

df_202123_all <- readRDS("~/Desktop/gl_nfs/Project_INTEGRATE/Analysis/Resistance_dynamics/2025-09-18_master_kpn_first_isolate_dataset_curation/results/CA_clinical_CRE_metadata_091825.RDS")  %>% subset(study == '2021-23')
```

# Coordinate ST data

```{r}
OLD_ST <- unique(df_201415$Species_ST) %>% sort
new_ST <- unique(df_202123$Species_ST) %>% sort 
ST_over_1 <- table(df$Species_ST) %>% subset(.>1) %>% names

df$Species_ST_new <- ifelse(df$Species_ST %in% c(ST_over_1),df$Species_ST,"Singleton STs") %>% gsub("Klebsiella pneumoniae ","",.)
                                          
break_of_sts <-  c(sort(unique(df$Species_ST_new %>% subset(.!="Singleton STs"))),"Singleton STs")
break_of_sts_renamed <- gsub("Klebsiella pneumoniae ","",break_of_sts)
       
ST_scale <- scale_fill_manual(breaks = break_of_sts,values = c(hues::iwanthue(n = length(unique(df$Species_ST_new)) -1),"lightgray"),labels = break_of_sts_renamed,name = "Sequence type", guide = guide_legend(ncol=7, title.position = "top", label.position = "right"))


df$Species_ST_new_no_species <- gsub("Klebsiella pneumoniae ","",df$Species_ST)
 
df_201415 <- df%>% subset(study == "2014-15")
df_202123<- df%>% subset(study == "2021-23")
```

```{r}
amrfinder <- readRDS("~/Desktop/gl_nfs/Project_INTEGRATE/Analysis/Resistance_dynamics/2025-09-18_master_kpn_first_isolate_dataset_curation/results/amrfinder_cleaned_091825.RDS")
amrfinder_beta <- amrfinder %>% subset(class == "BETA-LACTAM")

amrfinder_carb_subclass <- subset(amrfinder,subclass == "CARBAPENEM")
amrfinder_carb <- amrfinder %>% subset(subclass == "CARBAPENEM" | grepl("KPC|NDM|VIM|IMP",element_symbol) | grepl("OXA-48 family",element_name))
amrfinder_matrix <- readRDS("~/Desktop/gl_nfs/Project_INTEGRATE/Analysis/Resistance_dynamics/2025-09-18_master_kpn_first_isolate_dataset_curation/results/amrfinder_matrix_091825.RDS")
amrfinder_matrix_beta <- amrfinder_matrix %>% select(any_of(sort(unique(amrfinder_beta$element_symbol))))
amrfinder_matrix_carb <- amrfinder_matrix %>% select(any_of(sort(unique(amrfinder_carb$element_symbol))))

# Melt dataframe matrix so that 
isolate_level_data <- amrfinder_matrix  %>% mutate(isolate_no = rownames(.)) %>% reshape2::melt(id.vars="isolate_no")
isolate_level_data <- left_join(isolate_level_data, df %>% select(isolate_no,Species,Species_ST,study,hospital_location,rectal), by = "isolate_no") 

# Get an isolate's carbapenemase genes
df$blactamase <- sapply(df$isolate_no,FUN=function(x){
  amrfinder_beta %>% subset(isolate_no == x) %>% .$element_symbol %>% paste0(collapse=";")
}) %>% {ifelse(.=="","No blactamase",.)}


df$carbapenemase <- sapply(df$isolate_no,FUN=function(x){
  amrfinder_carb %>% subset(isolate_no == x) %>% .$element_symbol %>% paste0(collapse=";")
}) %>% {ifelse(.=="","No carbapenemase",.)}

df$carbapenemase_family <- sapply(df$isolate_no,FUN=function(x){
    carbapenemase <- amrfinder_carb %>% subset(isolate_no == x) %>% .$element_symbol
     if(is_empty(carbapenemase)){
      "No carbapenemase"
    } else {
      sapply(carbapenemase,FUN=function(y){
        str_split(y,"-",simplify=T) %>% .[,1]
    }) %>% unlist  %>% paste0(collapse=";") 
    }
    
})  
df$blaKPC_presence <- grepl("KPC",df$carbapenemase_family) 
df$blaKPC_2 <- grepl("KPC-2",df$carbapenemase)  & !grepl("KPC-29",df$carbapenemase)
df$no_carbapenemase <- grepl("No carbapenemase",df$carbapenemase_family) 
df_201415 <- df%>% subset(study == "2014-15")
df_202123<- df%>% subset(study == "2021-23")
```

# Kleborate
```{r}
kleborate <- readRDS("~/Desktop/gl_nfs/Project_INTEGRATE/Analysis/Resistance_dynamics/2025-09-18_master_kpn_first_isolate_dataset_curation/results/kleborate_both_studies_091825.RDS") %>% subset(species == "Klebsiella pneumoniae") %>% subset(isolate_no %in% df$isolate_no)
rownames(kleborate) <- kleborate$isolate_no
kleborate_matrix <- gontjesr::get_presence_absence_matrix(kleborate,'Omp_mutations') %>% as.data.frame
kleborate_matrix$OmpK36_mutation <- kleborate_matrix %>% as.data.frame %>% select_if(grepl("OmpK36",colnames(.))) %>% rowSums() %>% {ifelse(.>0,1,0)}
kleborate_matrix$OmpK35_mutation <- kleborate_matrix %>% as.data.frame %>% select_if(grepl("OmpK35",colnames(.))) %>% rowSums() %>% {ifelse(.>0,1,0)}
kleborate_matrix$porin_mutation <-  kleborate_matrix$OmpK36_mutation  == 1 | kleborate_matrix$OmpK35_mutation ==1

df <-left_join(df,kleborate_matrix %>% mutate(isolate_no = rownames(.))) %>% left_join(amrfinder_matrix_beta %>% mutate(isolate_no = rownames(.)))
```

# Differences across study periods in carbapenemase
```{r}
convert_tableone_into_df(dataset = df,vars = c('blaKPC_presence','carbapenemase',"porin_mutation","OmpK35_mutation","OmpK36_mutation"),strata = 'study',outcome_names = c("2014-15","2021-23"),factorVars = c('carbapenemase',"porin_mutation","OmpK35_mutation","OmpK36_mutation")) %>% favorite_kable()
```

# Differences among noncarbepenmase genomes
```{r}
df_nocarb <- df %>% subset(carbapenemase == 'No carbapenemase')
convert_tableone_into_df(dataset = df_nocarb,vars = c('blaKPC_presence','carbapenemase',"porin_mutation","OmpK35_mutation","OmpK36_mutation"),strata = 'study',outcome_names = c("2014-15","2021-23"),factorVars = c('carbapenemase',"porin_mutation","OmpK35_mutation","OmpK36_mutation")) %>% favorite_kable()
```

# Data across ST258
```{r}
df_ST258 <- subset(df,df$ST == "258")
df_ST258$GD_KPC2 <- df_ST258$OmpK36GD==1 & df_ST258$`blaKPC-2` ==1
convert_tableone_into_df(dataset = df_ST258,vars = c('blaKPC_presence','carbapenemase',"porin_mutation","OmpK35_mutation","OmpK36_mutation","OmpK36GD","OmpK36TD","GD_KPC2"),strata = 'study',outcome_names = c("2014-15","2021-23"),factorVars = c('carbapenemase',"porin_mutation","OmpK35_mutation","OmpK36_mutation","OmpK36GD","OmpK36TD","GD_KPC2")) %>% favorite_kable()
```

# Mobsuite data
## INTEGRATE
```{r}  
mobsuite_contig <- read_delim("~/Desktop/gl_nfs/Project_INTEGRATE/Analysis/Descriptive_epidemiology/2024-12-30_mobsuite/results/INTEGRATE_contig_report.txt")
mobsuite_contig$isolate_no <- gsub(".fna","",mobsuite_contig$sample_id) 
mobsuite_contig <- subset(mobsuite_contig,isolate_no %in% df$isolate_no)
mobtyper <- read_delim("~/Desktop/gl_nfs/Project_INTEGRATE/Analysis/Descriptive_epidemiology/2024-12-30_mobsuite/results/INTEGRATE_mobtyper_results.txt")
mobtyper$isolate_no <-mobtyper$sample_id %>% str_split(.,":",simplify=T) %>% .[,1] %>%  gsub(".fna","",.)
mobtyper <- subset(mobtyper,isolate_no %in% df$isolate_no)
```
## 2014-15 study
```{r}
 amrfinder_mat <- subset(amrfinder_matrix,rownames(amrfinder_matrix) %in% df$isolate_no)
amrfinder <- readRDS("~/Desktop/gl_nfs/Project_INTEGRATE/Analysis/Resistance_dynamics/2025-09-18_master_kpn_first_isolate_dataset_curation/results/amrfinder_cleaned_091825.RDS")
amrfinder_carb <- subset(amrfinder,subclass == "CARBAPENEM") %>% subset(isolate_no %in% df$isolate_no)
mobtyper_old <-  read_delim("~/Desktop/gl_nfs/Project_Penn_KPC/Analysis/Combination_resistance/pre-intro-blbli-resistance-ms/data/mobsuite/CRKP_mobtyper_results.txt")
mobtyper_old$isolate_no <-mobtyper_old$sample_id %>% str_split(.,":",simplify=T) %>% .[,1] %>%  gsub("_contigs_l1000","",.)
mobtyper_old <- subset(mobtyper_old,isolate_no %in% df$isolate_no)
mobsuite_contig_old <-  read_delim("~/Desktop/gl_nfs/Project_Penn_KPC/Analysis/Combination_resistance/pre-intro-blbli-resistance-ms/data/mobsuite/CRKP_contig_report.txt")
mobsuite_contig_old$isolate_no <-mobsuite_contig_old$sample_id %>% str_split(.,":",simplify=T) %>% .[,1] %>%  gsub("_contigs_l1000","",.)
mobsuite_contig_old <- subset(mobsuite_contig_old,isolate_no %in% df$isolate_no)

mobtyper_merged <- bind_rows(mobtyper_old %>% select(isolate_no,num_contigs,size,gc,primary_cluster_id,mash_nearest_neighbor),
                             mobtyper %>% select(isolate_no,num_contigs,size,gc,primary_cluster_id,mash_nearest_neighbor)) %>% mutate(study =ifelse( grepl("INT",isolate_no),"2021-23",'2014-15'))
mobsuite_contig_merged <- bind_rows(mobsuite_contig_old %>% select(isolate_no,contig_id,size,gc,primary_cluster_id,mash_nearest_neighbor),
                             mobsuite_contig %>% select(isolate_no,contig_id,size,gc,primary_cluster_id,mash_nearest_neighbor))%>% mutate(study =ifelse( grepl("INT",isolate_no),"2021-23",'2014-15'))
mobtyper_mat <- gontjesr::get_mobtyper_cluster_matrix(df,mobtyper_merged)
```

KPC plasmid matrix
```{r}
get_KPC_plasmid <- function(isolate,amrfinder_carb,mobsuite_contig){
  amrfinder_sub <- amrfinder_carb %>% subset(isolate_no == isolate)
  mobsuite_contig_merged <- mobsuite_contig_merged %>% subset(isolate_no == isolate & contig_id %in% amrfinder_sub$contig_id)
  mobsuite_contig_merged$primary_cluster_id %>% paste0(collapse=';') %>% {ifelse(.=="","No carbapenemase",.)} %>% {ifelse(.=="-","Chromosome",.)}
}


df$KPC_plasmid <- lapply(df$isolate_no,get_KPC_plasmid,amrfinder_carb,mobsuite_contig) %>% unlist

mobtyper_merged$KPC_plasmid <- apply(mobtyper_merged,1,FUN=function(x){
  isolate_kpc <- df %>% subset(isolate_no == x[['isolate_no']]) %>% .$KPC_plasmid %>% str_split(.,";" ) %>% unlist
  sum(x[['primary_cluster_id']] %in% isolate_kpc) 
})

mobtyper_merged$carbapenemase <-apply(mobtyper_merged,1,FUN=function(x){
  mobsuite_contigs <- mobsuite_contig_merged %>% subset(primary_cluster_id == x[["primary_cluster_id"]] & isolate_no == x[["isolate_no"]] ) %>% .$contig_id
  amrfinder_sub <- amrfinder_carb %>% subset(contig_id %in% mobsuite_contigs)
  if(nrow(amrfinder_sub)==0){
    ""
  } else{
    
  paste0(amrfinder_sub$element_symbol_renamed,collapse=';')
  }
   
})

df_202123<- df%>% subset(study == "2021-23")

# Convert KPC_plasmid variable into a matrix where row is isolate and entry is 1 for present and 0 for absent
KPC_plasmid_matrix <- df_202123 %>% select(isolate_no,KPC_plasmid) %>% 
    pivot_wider(names_from = KPC_plasmid, values_from = KPC_plasmid) %>%     column_to_rownames("isolate_no")  %>% {ifelse(is.na(.),0,1)}  %>% as.data.frame
```

# Descriptive statistics
## All
```{r}
paste0("Total plasmids: ",ncol(mobtyper_mat))
old_plasmids <- mobtyper_mat %>% subset(grepl("PCMP",rownames(.))) %>% colSums(.) %>% subset(.>0) %>% names
new_plasmids <- mobtyper_mat %>% subset(grepl("INT",rownames(.))) %>% colSums(.) %>% subset(.>0) %>% names
ggVennDiagram::ggVennDiagram(list(old_plasmids = old_plasmids, new_plasmids = new_plasmids))

```

## KPC
```{r}
paste0("Total KPC plasmids: ",length(unique(df$KPC_plasmid %>% subset(!. %in% c("Chromosome","No carbapenemase")))))

df$AA018 <- ifelse(df$KPC_plasmid =="AA018",T,F)
df$AA038 <- ifelse(df$KPC_plasmid %in% c("AA038",""),T,F)
df$AA274 <- ifelse(df$KPC_plasmid =="AA274",1,0)
df$AD548 <- ifelse(df$KPC_plasmid =="AD548",1,0)
df$AA356 <- ifelse(df$KPC_plasmid =="AA356",1,0) 
df$AA552 <- ifelse(df$KPC_plasmid =="AA552",1,0)
df$AA357 <- ifelse(df$KPC_plasmid =="AA357",1,0)
df$AA275 <- ifelse(df$KPC_plasmid =="AA275",1,0)
df$AA038 <- ifelse(df$KPC_plasmid =="AA038",1,0)
df$AA434 <- ifelse(df$KPC_plasmid =="AA434",1,0)

KPC_plasmids_old <- unique(df %>% subset(study == '2014-15') %>% .$KPC_plasmid %>% str_split(.,";") %>% unlist) %>% subset(!. %in% c("Chromosome","No carbapenemase"))
KPC_plasmids_new <- unique(df %>% subset(study == '2021-23') %>%  .$KPC_plasmid %>% str_split(.,";") %>% unlist)%>% subset(!. %in% c("Chromosome","No carbapenemase"))

ggVennDiagram::ggVennDiagram(list(KPC_plasmids_old = KPC_plasmids_old, KPC_plasmids_new = KPC_plasmids_new))

variables <-  c("KPC_plasmid","AA018","AA274",'AD548','AA356','AA552','AA357','AA275','AA038','AA434')

gontjesr::convert_tableone_into_df(dataset=df %>% subset(KPC_plasmid != "No carbapenemase"),vars =variables,strata='study',outcome_names = c("2014-15",'2021-23'),factorVars =variables) %>% favorite_kable()
```

### ST258
```{r}
paste0("Total KPC plasmids: ",length(unique(df %>% subset(ST=='258') %>% .$KPC_plasmid %>% subset(!. %in% c("Chromosome","No carbapenemase")))))

KPC_plasmids_old <- unique(df %>% subset(study == '2014-15' & ST=='258') %>% .$KPC_plasmid %>% str_split(.,";") %>% unlist) %>% subset(!. %in% c("Chromosome","No carbapenemase"))
KPC_plasmids_new <- unique(df %>% subset(study == '2021-23' & ST=='258') %>%  .$KPC_plasmid %>% str_split(.,";") %>% unlist)%>% subset(!. %in% c("Chromosome","No carbapenemase"))

ggVennDiagram::ggVennDiagram(list(KPC_plasmids_old = KPC_plasmids_old, KPC_plasmids_new = KPC_plasmids_new))

gontjesr::convert_tableone_into_df(dataset=df %>% subset(ST=='258' & KPC_plasmid != "No carbapenemase"),vars = variables,strata='study',outcome_names = c("2014-15",'2021-23'),argsNormal = 'KPC_plasmid',factorVars = variables) %>% favorite_kable()
```

